dev/src/reference.ts

```diff
@@ -700,14 +701,118 @@ export class FieldOrder {
   }
 }
 
+abstract class FilterInternal {
+  /** Returns a list of all field filters that are contained within this filter */
+  abstract getFlattenedFilters(): FieldFilterInternal[];
+
+  /** Returns a list of all filters that are contained within this filter */
+  abstract getFilters(): FilterInternal[];
+
+  /** Returns the field of the first filter that's an inequality, or null if none. */
+  abstract getFirstInequalityField(): FieldPath | null;
+
+  /** Returns the proto representation of this filter */
+  abstract toProto(): Filter;
+
+  abstract isEqual(other: FilterInternal): boolean;
+}
+
+class CompositeFilterInternal extends FilterInternal {
+  constructor(
+    private filters: FilterInternal[],
+    private operator: api.StructuredQuery.CompositeFilter.Operator
+  ) {
+    super();
+  }
+
+  // Memoized list of all field filters that can be found by traversing the tree of filters
+  // contained in this composite filter.
+  private memoizedFlattenedFilters: FieldFilterInternal[] | null = null;
+
+  public getFilters(): FilterInternal[] {
+    return this.filters;
+  }
+
+  public getFirstInequalityField(): FieldPath | null {
+    return (
+      this.getFlattenedFilters().find(filter => filter.isInequalityFilter())
+        ?.field ?? null
+    );
+  }
+
+  public isConjunction(): boolean {
+    return this.operator === 'AND';
+  }
+
+  public getFlattenedFilters(): FieldFilterInternal[] {
+    if (this.memoizedFlattenedFilters !== null) {
+      return this.memoizedFlattenedFilters;
+    }
+
+    this.memoizedFlattenedFilters = this.filters.reduce(
+      (allFilters: FieldFilterInternal[], subfilter: FilterInternal) =>
+        allFilters.concat(subfilter.getFlattenedFilters()),
+      []
+    );
+
+    return this.memoizedFlattenedFilters;
+  }
+
+  public toProto(): api.StructuredQuery.IFilter {
+    if (this.filters.length === 1) {
```

Comment:
Just a note here that for the datastore PR, we still use an `AND` filter with one element because it fits in naturally with the way that the code already is.